File: src/main/java/ru/nucodelabs/algorithms/forward_solver/ForwardSolver.java
package ru.nucodelabs.algorithms.forward_solver;


import ru.nucodelabs.data.ves.ExperimentalData;
import ru.nucodelabs.data.ves.ModelLayer;

import java.util.List;

public interface ForwardSolver {

    /**
     * Returns default ForwardSolver implementation
     *
     * @param experimentalData experimental data of picket
     * @param modelData        model data of picket
     * @return new instance
     */
    static ForwardSolver createDefaultForwardSolver(List<ExperimentalData> experimentalData, List<ModelLayer> modelData) {
        return createSonetForwardSolver(experimentalData, modelData);
    }

    /**
     * Returns Sonet implementation of ForwardSolver
     *
     * @param experimentalData experimental data of picket
     * @param modelData        model data of picket
     * @return new instance
     */
    static ForwardSolver createSonetForwardSolver(List<ExperimentalData> experimentalData, List<ModelLayer> modelData) {
        return new SonetForwardSolver(experimentalData, modelData);
    }

    /**
     * Get solved resistance_apparent for chart
     *
     * @return resistance apparent
     */
    List<Double> solve();
}

File: src/main/java/ru/nucodelabs/algorithms/forward_solver/ForwardSolver.kt
package ru.nucodelabs.algorithms.forward_solver

import ru.nucodelabs.data.ves.ExperimentalData
import ru.nucodelabs.data.ves.ModelLayer
import ru.nucodelabs.mathves.SonetForwardSolver

interface ForwardSolver {
    /**
     * Returns solved resistance values for `modelData` that match to distances of `experimentalData`
     */
    operator fun invoke(experimentalData: List<ExperimentalData>, modelData: List<ModelLayer>): List<Double>

    companion object Factory {
        @JvmStatic
        fun createDefault(): ForwardSolver = ForwardSolver()
    }
}

fun ForwardSolver(): ForwardSolver = object : ForwardSolver {
    private val delegate = SonetForwardSolver()
    override fun invoke(experimentalData: List<ExperimentalData>, modelData: List<ModelLayer>): List<Double> {
        return delegate.solve(
            modelData.map { it.resistance }.toDoubleArray(),
            modelData.map { it.power }.toDoubleArray(),
            modelData.size,
            experimentalData.map { it.ab2 }.toDoubleArray(),
            experimentalData.map { it.mn2 }.toDoubleArray(),
            experimentalData.size
        ).toList()
    }
}